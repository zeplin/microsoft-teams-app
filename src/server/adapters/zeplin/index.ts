export * from "./zeplin";
export * from "./types";
