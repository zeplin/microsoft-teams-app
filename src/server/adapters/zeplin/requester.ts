import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { ZeplinError } from "./ZeplinError";
import { INTERNAL_SERVER_ERROR } from "http-status-codes";

export class Requester {
    private instance: AxiosInstance;
    constructor(config: AxiosRequestConfig) {
        this.instance = axios.create(config);
    }

    async delete(url: string, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.delete(url, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async get<T = object>(url: string, config?: AxiosRequestConfig): Promise<T> {
        try {
            const { data } = await this.instance.get<T>(url, config);
            return data;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async patch(url: string, data: object, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.patch(url, data, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async post(url: string, data: object, config?: AxiosRequestConfig): Promise<string> {
        try {
            const { data: { id } } = await this.instance.post<{ id: string }>(url, data, config);
            return id;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }
}
