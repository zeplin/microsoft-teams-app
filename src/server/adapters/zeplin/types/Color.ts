export interface Color {
    id: string;
    created: number;
    name: string;
    r: number;
    g: number;
    b: number;
    a: number;
}
