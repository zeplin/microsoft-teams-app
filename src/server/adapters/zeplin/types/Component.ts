import { Thumbnails } from "./Thumbnails";

export interface Component {
    id: string;
    name: string;
    created: number;
    updated?: number;
    description?: string;
    image?: {
        width: number;
        height: number;
        original_url?: string;
        thumbnails?: Thumbnails;
    };
    section?: {
        id: string;
        group?: {
            id: string;
        };
    };
}
