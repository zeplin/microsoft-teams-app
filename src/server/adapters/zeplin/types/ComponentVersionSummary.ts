import { User } from "./User";
import { Thumbnails } from "./Thumbnails";

enum DestinationType {
    SCREEN = "screen",
    PREVIOUS = "previous",
}

export interface ComponentVersionSummary {
    id: string;
    created: number;
    creator?: User;
    commit?: {
        color: {
            r: number;
            g: number;
            b: number;
            a: number;
        };
        message?: string;
        author?: User;
    };
    image_url?: string;
    thumbnails?: Thumbnails;
    background_color?: {
        r: number;
        g: number;
        b: number;
        a: number;
    };
    source: string;
    width: number;
    height: number;
    density_scale: number;
    links: {
        rect: {
            x: number;
            y: number;
            width: number;
            height: number;
        };
        destination: {
            name?: string;
            type: DestinationType;
        };
    }[];
}
