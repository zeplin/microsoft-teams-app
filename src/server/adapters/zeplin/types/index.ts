export * from "./Organization";
export * from "./Project";
export * from "./Styleguide";
export * from "./RemPreferences";
export * from "./WebhookEvents";
export * from "./WebhookResourceType";
export * from "./ZeplinApplication";
export * from "./ProjectWebhook";
export * from "./StyleguideWebhook";
export * from "./User";
export * from "./WebhookStatus";
export * from "./WebhookUrlHealth";
