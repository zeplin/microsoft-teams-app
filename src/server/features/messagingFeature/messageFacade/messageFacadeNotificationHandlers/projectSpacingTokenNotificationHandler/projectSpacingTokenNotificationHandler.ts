import {
    WebhookEvent,
    EventPayload,
    ProjectContext,
    EventType
} from "../../../messagingTypes";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, AdaptiveCard } from "../teamsCardTemplates";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../../config";
import { URL } from "url";
import { SpacingTokenResource } from "../resources";

type ProjectSpacingTokenEventDescriptor = {
    type: EventType.PROJECT_SPACING_TOKEN;
    action: "created" | "updated";
};

class ProjectSpacingTokenNotificationHandler extends NotificationHandler {
    delay = SHORT_DELAY;
    private getText(events: WebhookEvent<ProjectSpacingTokenEventPayload>[]): string {
        const [{
            payload: {
                action,
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        name: pivotSpacingTokenName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added" : "updated";
        return events.length === 1
            ? `**${pivotSpacingTokenName}** is ${actionText} in _${projectName}_! üèÉ‚Äç‚ôÇ`
            : `**${events.length} spacing tokens** are ${actionText} in _${projectName}_! üèÉ‚Äç‚ôÇ`;
    }

    private getWebappURL(
        events: WebhookEvent<ProjectSpacingTokenEventPayload>[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `project/${projectId}/styleguide/spacing`;
        events.forEach(event => webappURL.searchParams.append("sptid", event.payload.resource.id));
        return webappURL.toString();
    }

    private getMacAppURL(
        events: WebhookEvent<ProjectSpacingTokenEventPayload>[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        return `${ZEPLIN_MAC_APP_URL_SCHEME}spacing?pid=${projectId}&sptids=${events.map(event => event.payload.resource.id).join(",")}`;
    }

    shouldHandleEvent(event: WebhookEvent): boolean {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: WebhookEvent<ProjectSpacingTokenEventPayload>[]
    ): AdaptiveCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: "Make sure your stylesheets are up to date!"
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export type ProjectSpacingTokenEventPayload = EventPayload<
    ProjectSpacingTokenEventDescriptor,
    ProjectContext,
    SpacingTokenResource
>;
export const projectSpacingTokenNotificationHandler = new ProjectSpacingTokenNotificationHandler();