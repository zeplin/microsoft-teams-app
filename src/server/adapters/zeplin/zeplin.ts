import { ProjectWebhooks } from "./ProjectWebhooks";
import { Requester } from "./requester";
import { StyleguideWebhooks } from "./StyleguideWebhooks";

interface ZeplinOptions {
    url: string;
    webhookSecret: string;
}

class Zeplin {
    public projectWebhooks!: ProjectWebhooks;
    public styleguideWebhooks!: StyleguideWebhooks;

    init({ url, webhookSecret }: ZeplinOptions): void {
        const requester = new Requester({ baseURL: `${url}/v1/` });
        this.projectWebhooks = new ProjectWebhooks(requester, webhookSecret);
        this.styleguideWebhooks = new StyleguideWebhooks(requester, webhookSecret);
    }
}

export const zeplin = new Zeplin();
