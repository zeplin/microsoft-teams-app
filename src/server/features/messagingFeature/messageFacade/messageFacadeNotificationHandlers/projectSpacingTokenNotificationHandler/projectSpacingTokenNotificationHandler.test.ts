import { projectSpacingTokenNotificationHandler } from "./projectSpacingTokenNotificationHandler";
import { ProjectSpacingTokenEvent } from "../../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    spacingTokenId?: string;
    spacingTokenName?: string;
}

function getDummyEvent({
    action = "created",
    spacingTokenId = "spacingTokenId",
    spacingTokenName = "spacingTokenName"
}: GetDummyEventParams = {}): ProjectSpacingTokenEvent {
    return {
        payload: {
            action,
            context: {
                project: {
                    id: "projectId",
                    name: "projectName"
                }
            },
            resource: {
                id: spacingTokenId,
                data: {
                    id: spacingTokenId,
                    name: spacingTokenName
                }
            }
        }
    } as ProjectSpacingTokenEvent;
}

describe("projectSpacingTokenNotificationHandler", () => {
    describe("getTeamsMessage method", () => {
        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectSpacingTokenNotificationHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectSpacingTokenNotificationHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            spacingTokenId: "anotherSpacingTokenId",
                            spacingTokenName: "anotherSpacingTokenName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectSpacingTokenNotificationHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectSpacingTokenNotificationHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            spacingTokenId: "anotherSpacingTokenId",
                            spacingTokenName: "anotherSpacingTokenName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });
    });
});
