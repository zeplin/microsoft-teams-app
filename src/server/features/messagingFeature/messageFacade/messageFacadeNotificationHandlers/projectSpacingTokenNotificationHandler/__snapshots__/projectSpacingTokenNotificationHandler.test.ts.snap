// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectSpacingTokenNotificationHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 spacing tokens** are added in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%2CanotherSpacingTokenId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&sptid=anotherSpacingTokenId"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenNotificationHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenNotificationHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 spacing tokens** are updated in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%2CanotherSpacingTokenId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&sptid=anotherSpacingTokenId"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenNotificationHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! üèÉ‚Äç‚ôÇ",
  "sections": [
    {
      "text": "Make sure your stylesheets are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId"
        }
      ]
    }
  ]
}
`;
